src/reporter/index.js

```diff
@@ -89,13 +89,13 @@ export default class Reporter {
             reportItem     = this.reportQueue.shift();
             currentFixture = reportItem.fixture;
 
-            await this.plugin.reportTestDone(reportItem.test.name, reportItem.testRunInfo, reportItem.test.meta);
-
             // NOTE: here we assume that tests are sorted by fixture.
             // Therefore, if the next report item has a different
             // fixture, we can report this fixture start.
             nextReportItem = this.reportQueue[0];
 
+            await this.plugin.reportTestDone(reportItem.test.name, reportItem.testRunInfo, reportItem.test.meta);
```

Comment:
At present, we raise the `reportFixtureStart` event after `reportTestDone` event (by some conditions) in the `Reporter` class.

I propose to use another algorithm.
1)Raise an `fixture-start` event inside the `BrowserJob` class and pass it via `BrowserJob`-> `Task` -> `Reporter` chain.
2)`Reporter` class just handles this event.
3)Calculate the first test of each fixture [here](https://github.com/DevExpress/testcafe/blob/master/src/runner/browser-job.js#L25)
4)Raise a `fixture-start` event before emitting the `'test-run-start'` if the current test is the first test of fixture
https://github.com/DevExpress/testcafe/blob/5b4979d60ced70c54e7e331847c60d90ec8be3cb/src/runner/browser-job.js#L42

E.g. `reportFixtureStart` event will be raised before `reportTestStart` event.

@AndreyBelym What do you think about it?

Reply:
If we raise the `reportFixtureStart` event inside the `BrowserJob` class it'll require to add some extra logic related to running tests and emit `fixture-start` event in multiple browsers. At this moment this logic is based on the `reportItem.pendingRuns` field where `pendingRuns` is equal to number of browsers. However, `BrowserJob` is responsible only for one browser, so I assume that it would be better to leave this synchronization mechanism as is.

Reply:
Ok. Then you can use only second part of the previos comment:
* calculate the `test.isFirstTestOfFixture` property for each test after bootstrapping
* move the `plugin.reportFixtureStart` method call before the `plugin.reportTestStart` method call.

```js
if (!reportItem.pendingStarts) {
    await this.plugin.reportFixtureStart(...);

     if (this.plugin.reportTestStart)
         await this.plugin.reportTestStart(reportItem.test.name, reportItem.test.meta);

    reportItem.pendingTestRunStartPromise.resolve();
}
```

Reply:
I thought about this approach after the discussion. However, in this case `reportTestDone` of the previous test can be raised after the `reportFixtureStart` of the current test. 

```
Connected Firefox 72.0 / Windows 10
Connected Internet Explorer 11.0 / Windows 10
Connected Chrome 80.0.3987.122 / Windows 10
  [Regression](GH-4787) - Should wait for last report before new fixture starts
Fixture 1
start: Test 1
start: Test 1
start: Test 1
done: Test 1
start: Test 1
done: Test 1
done: Test 1
start: Test 1
done: Test 1
start: Test 1
done: Test 1
start: Test 1
done: Test 1
start: Test 1
done: Test 1
start: Test 1
done: Test 1
start: Test 1
done: Test 1
**Fixture 2
done: Test 1**
start: Test 2
start: Test 2.1
done: Test 2.1
start: Test 2
done: Test 2
start: Test 2
done: Test 2
start: Test 2
done: Test 2
start: Test 2
done: Test 2
start: Test 2
done: Test 2
start: Test 2
done: Test 2
start: Test 2
done: Test 2
start: Test 2
done: Test 2
start: Test 2
done: Test 2
**Fixture 3**
start: Test 3.1
**done: Test 2**
start: Test 3
done: Test 3.1
start: Test 3
done: Test 3
start: Test 3
done: Test 3
start: Test 3
done: Test 3
start: Test 3
done: Test 3
start: Test 3
done: Test 3
start: Test 3
done: Test 3
start: Test 3
done: Test 3
start: Test 3
done: Test 3
start: Test 3
done: Test 3
done: Test 3
    √ Should wait for last report before new fixture starts (63576ms)
```